### PR TITLE
PSBT(bip174) implementation #4615

### PR DESCRIPTION
**Pull request by asfin**
Support for PSBT serialization format.
Unittests and integration tests provided.
Manually tested(bitcoin testnet) with UI, send to p2sh, p2pkh, p2wsh, p2wpkh. Export to file/qr/buffer works ok.
Resulting PSBT valid for bitcoind `walletprocesspsbt` API, and can be processed and finalized by it.
Fixes at least this issues #4615 #4514 #4471

About code:
I think package bip174.py should be renamed, because it doesn't contain BIP174 class. 
@SomberNight bip32.py refactoring was conflicted with mine xpubkey_utils.py, so i removed xpubkey_utils.py and just moved rest of it to bip32.py
HW plugins wasn't tested, so if you saw an error, write about it(or just fix).
Some parts of code wasn't properly finalized, because i think that review and some time after merge should find some problems.

### CODE REVIEW

File: electrum/wallet.py [hunk 1]
```diff
@@ -1521,13 +1864,25 @@ def create_new_address(self, for_change=False):
...
             self._addr_to_addr_index[address] = (for_change, n)
+            self._addr_to_addr_path[address] = "/{:d}/{:d}".format(for_change, n)
             self.save_addresses()
             self.add_address(address)
             if for_change:
                 # note: if it's actually used, it will get filtered later
                 self._unused_change_addresses.append(address)
             return address
 
+    def create_new_address_by_path(self, derivation_path: str, password):
```
[the remark this answers is not in the document]
Comment (asfin) [reply electrum/wallet.py]: in old tests, tried to reproduce original test vectors, but without success, i think this should be removed

File: electrum/wallet.py [hunk 2]
```diff
@@ -1507,10 +1847,13 @@ def min_acceptable_gap(self):
     def load_addresses(self):
         super().load_addresses()
         self._addr_to_addr_index = {}  # key: address, value: (is_change, index)
+        self._addr_to_addr_path = {}   # key: address, value: /0/0'/2
```
[the remark this answers is not in the document]
Comment (asfin) [reply electrum/wallet.py]: bip32_derivation struct needs full path, and addr_to_addr_index doesn't have enough data if derivation level 3+ used.

File: electrum/wallet.py [hunk 3]
```diff
@@ -592,8 +640,7 @@ def get_unconfirmed_base_tx_for_batching(self) -> Optional[Transaction]:
             return tx
         return candidate
 
-    def make_unsigned_transaction(self, coins, outputs, config, fixed_fee=None,
-                                  change_addr=None, is_sweep=False):
+    def make_unsigned_transaction(self, coins, outputs, config, fixed_fee=None, change_addr=None):
```
[the remark this answers is not in the document]
Comment (asfin) [reply electrum/wallet.py]: @SomberNight removed together with mktx

**Comment by asfin** [comment]
Removed all mktx related stuff. 
Added support for PSBT to Old Keystore. This is covered by `test_sending_between_p2sh_2of3_and_uncompressed_p2pkh`, `test_old_keystore_provides_valid_bip32_keys`
Offline tests are broken, trying to figure out possibility to fix it.

**Comment by SomberNight** [comment]
> But is it good behavior to populate history from partial transactions?

These txns show up in the `History` tab, right? It might be a bit confusing... but not sure how to solve it otherwise atm.

Where in the code do you specifically need the input/parent transactions? I wonder what side effects of `add_transaction` (called by `receive_tx_callback`) are needed. Would this be enough: `self.transactions[tx_hash] = tx` instead?

**Comment by asfin** [comment]
Main purpose - `is_mine` wallet check, which triggers populate and process logic(check amounts, get associated keys for utxo, set `txin['type']`). And simply using `self.transactions[tx_hash] = tx` doesn't work.
In case of populating history with this transactions, may it broke transaction synchronization? Because if synchronization performs only on last timestamps/block_count, it may shadow transactions which was made before previous transaction and current.

File: electrum/transaction.py [hunk 4]
```diff
@@ -1262,6 +412,417 @@ def as_dict(self):
...
+    """
+
+    def update(self, raw):
+        raise NotImplementedError('Immutable object')
+
+    def input_script(self, txin: dict, estimate_size=False):
+        script_sig = txin.get('scriptSig', '')
+        return script_sig
+
+    def get_witness(self, i: int) -> str:
```
Comment (SomberNight) [inline electrum/transaction.py]: this is very confusingly named.
it returns a serialized output (amount + scriptPubKey)...
How about `get_serialized_output` maybe?

File: electrum/transaction.py [hunk 5]
```diff
@@ -1262,6 +412,417 @@ def as_dict(self):
...
+        txo = self.outputs()[i]
+        witness = serialize_output(txo)
+        if witness == '':
+            witness = '00'
```
Comment (SomberNight) [inline electrum/transaction.py]: Is this branch ever entered?

File: electrum/wallet.py [hunk 6]
```diff
@@ -814,16 +1011,120 @@ def add_input_info(self, txin):
...
         if tx.is_complete():
             return False
         # add info to inputs if we can; otherwise we might return a false negative:
-        tx.add_inputs_info(self)
+        if isinstance(tx, Transaction):
+            self.add_input_info_to_all_inputs(tx)  # though note that this is a side-effect
+        else:
+            self.add_info_to_psbt(tx)
+            if self._can_sign(tx):
+                return True
```
Comment (SomberNight) [inline electrum/wallet.py]: watch-only wallets will get here as `_can_sign` seems not to be finished.
so e.g. xpub only wallet can click "sign" button in tx dialog

**Comment by SomberNight** [comment]
offline signing with an old electrum seed (`OldKeystore`; discussed recently on IRC) does not work
the produced transactions fails validation

I wrote a test for this in https://github.com/spesmilo/electrum/commit/e8a8a172174ff49adff556be6b025fc6853b20ad; try to see why that fails

File: electrum/transaction.py [hunk 7]
```diff
@@ -1262,6 +412,417 @@ def as_dict(self):
...
+    reimplementing some methods in order to get necessary data from parent psbt object
+
+    More programmer friendly constructor, because we don't have to use explicit raw format(this work done in PSBT class)
+    """
+
+    def __init__(
+            self,
+            inputs: Sequence[dict],
+            outputs: Sequence[TxOutput],
+            version: int = 2,
```
[the remark this answers is not in the document]
Comment (asfin) [reply electrum/transaction.py]: bitcoind always uses version 2 in psbt scenarios. But i think it also should work with 1

File: electrum/bip32.py [hunk 8]
```diff
@@ -259,11 +288,105 @@ def bip32_public_derivation(xpub, branch, sequence):
...
+        raise BitcoinException("Cannot parse pubkey. prefix: {}".format(x_pubkey[0:2]))
+    if pubkey:
+        address = public_key_to_p2pkh(bfh(pubkey))
+    return pubkey, address
+
+
+def xpubkey_to_pubkey(x_pubkey: str):
+    pubkey, address = xpubkey_to_address(x_pubkey)
+    return pubkey
+
```
[the remark this answers is not in the document]
Comment (asfin) [reply electrum/bip32.py]: Inheritance it keystore class was confusing, i decided to move out utilities methods, to be more explicit. Move it back to `xpubkey_utils.py`

**Comment by asfin** [comment]
> Offline signing with an old electrum seed (OldKeystore; discussed recently on IRC) does not work.
The produced transaction fails validation.
Only difference, is compressed public key. I thought that pushing full pubkey shouldn't work. So, now there are full pubkeys, and test works.

File: electrum/wallet.py (hunk 6 above)
Comment (asfin) [reply electrum/wallet.py]: it's ok now?

File: electrum/transaction.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (SomberNight) [reply electrum/transaction.py]: Oh ok. Then let's use 2.

**Comment by asfin** [comment]
FYI, trezor test fails because import in trezorlib seems to be broken(at least in my environment)
https://github.com/trezor/python-trezor/blob/master/trezorlib/client.py#L21

**Comment by SomberNight** [comment]
> FYI, trezor test fails because import in trezorlib seems to be broken(at least in my environment)
https://github.com/trezor/python-trezor/blob/master/trezorlib/client.py#L21

is it importing mnemonic.py from electrum instead of the global mnemonic package? I've seen that before. pycharm is mangling the PYTHONPATH. maybe see https://github.com/spesmilo/electrum/issues/4907#issuecomment-445466838

**Comment by jonathancross** [comment]
Looks like there is some serious merge conflicts to resolve here.
@asfin ?

**Comment by SomberNight** [comment]
@asfin Thank you for your work. Master now has PSBT support with https://github.com/spesmilo/electrum/pull/5721

In the end we decided to take a different implementation approach, that IMHO makes the code easier to maintain going forward, mainly that
- there is only two Transaction classes (Transaction and PartialTransaction(Transaction)), where PartialTransaction fully conforms to the API of Transaction
  - further, Transactions have inputs: TxInput, and outputs: TxOutput; and PartialTransactions have inputs: PartialTxInput(TxInput), and outputs: PartialTxOutput(TxOutput); instead of the old ugly input dicts
- and we ripped the old electrum custom partial tx format out of the code

Given that the lightning branch was now merged into master, and mostly due to that this PR had lots of hard to fix conflicts, we decided not to try to rebase this and fix conflicts. This is in large part of course our fault, as this PR should have been reviewed and progressed earlier, before it had developed huge conflicts, but we were focusing on other things, mostly lightning. This PR had seemed a huge task to review and get merged, even at the beginning; and now that I have worked on https://github.com/spesmilo/electrum/pull/5721 for more than a month, I can only tell that indeed it's a lot of work even just to understand the implications and get this kind of change merged. See the design decisions section in the other PR; all of which was not actually discussed here.

Rest assured, your work was not wasted, as I had spent several full days just thinking about code architecture, and this PR was a good reference for some of the decisions.

**Pull request closed without merging**